[FEATURE] Utiliser uniquement la colonne user-logins.lastLoggedAt pour lire et écrire la date de dernière connexion (PIX-9010)

igorissen opened the pull request:
## :unicorn: Problème
Pour des soucis de retro-compatibilité, nous avons choisi de mettre à jour la même information, c'est-à-dire la date de dernière connexion d'un utilisateur, sur les tables suivantes `Users` et `UserLogins`.

A terme, nous voulons uniquement utiliser la table `UserLogins` pour porter l'information de la date de dernière connexion.

## :robot: Proposition
Retirer toutes utilisations de la méthode `userRepository.updateLastLoggedAt(...)`, et modifier le code existant pour n'utiliser que la propriété `lastLoggedAt` de la table `UserLogins`.

## :rainbow: Remarques
⚠️ Cette modification touche le scope de la team prescription. Il faudrait voir pour un scénario (ou plusieurs) pour s'assurer qu'aucune régression n'a été introduite dans leur scope.

## :100: Pour tester

Utiliser PixAdmin comme outil de contrôle pour vérifier que la date de dernière connexion est la bonne.

- Se connecter sur [PixAdmin](https://admin-pr6977.review.pix.fr/) avec le compte `superadmin@example.net`
- Une fois connecté sur Pix App avec un compte standard ou _partenaire_ (GAR, Pole Emploi, ...)
 - Faire une recherche avec l'ID ou le nom, prénom de l'utilisateur
 - Vérifier dans les détails de cet utilisateur que la date de dernière connexion est la bonne

### Scénarios possibles :

#### Connexion simple (identifiant ou e-mail, et mot de passe) sans date de dernière connexion

- Se connecter sur [PixApp](https://app-pr6977.review.pix.fr/) avec le compte `salvor.hardin`
- Vérifier, avec PixAdmin, sur les détails de ce compte qu'une date de dernière connexion est disponible

#### Connexion simple (identifiant ou e-mail, et mot de passe) avec date de dernière connexion

- Prendre note de la date de dernière connexion du compte avec PixAdmin
- Se connecter sur [PixApp](https://app-pr6977.review.pix.fr/) avec le compte `gaal.dornick`
- Vérifier, avec PixAdmin, sur les détails de ce compte à une nouvelle date de dernière connexion

#### Connexion GAR

Ce scénario est à tester en local avec l'utilisation du `Faux GAR`.

- Prendre note de la date de dernière connexion, si présente, du compte `Eliza Delajungle` avec PixAdmin
- Se connecter au `Faux GAR` avec ces informations
  - Saml ID = <id de votre choix>
  - Prénom = Eliza
  - Nom = Delajungle
- Fournir le code campagne `SCOBADGE1`
- Commencer le parcours
- Fournir la date de naissance du compte `01-01-2011`
- Une fois sur le parcours, tutoriel affiché ou page d'accueil affiché, vérifier sur PixAdmin la date de dernière connexion

#### Connexion OIDC (Pole Emploi, ...)

Ce scénario est à tester en local

- Prendre note de la date de dernière connexion, si présente, pour l'un des comptes que vous allez utiliser
- Se connecter via le SSO de Pole Emploi avec l'un des comptes fournis dans Confluence
- Créer un nouveau compte si ce n'est pas déjà le cas
- Vérifier sur PixAdmin que le compte a une nouvelle date de dernière connexion

igorissen committed:
refactor(api): modify authenticate-user usecase

use only userLoginRepository.updateLastLoggedAt

igorissen committed:
refactor(api): modify authenticate-external-user usecase

igorissen committed:
refactor(api): modify authenticate-oidc-user usecase

igorissen committed:
refactor(api): modify reconcile-oidc-user usecase

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr6977.review.pix.fr)
  * [App (.org)](https://app-pr6977.review.pix.org)
  * [Orga (.fr)](https://orga-pr6977.review.pix.fr)
  * [Orga (.org)](https://orga-pr6977.review.pix.org)
  * [Certif (.fr)](https://certif-pr6977.review.pix.fr)
  * [Certif (.org)](https://certif-pr6977.review.pix.org)
  * [1D](https://1d-pr6977.review.pix.fr)
  * [Admin](https://admin-pr6977.review.pix.fr)
  * [API](https://api-pr6977.review.pix.fr/api/)
  * [Audit Logger](https://pix-audit-logger-review-pr6977.osc-fr1.scalingo.io/api/)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr6977/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr6977/environment)
  * [scalingo audit-logger](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr6977/environment)

igorissen committed:
refactor(api): modify create-oidc-user usecase

igorissen committed:
refactor(api): modify create-user-and-reconcile-to-organization-learner-from-external-user usecase

igorissen committed:
refactor(api): modify get-external-authentication-redirection-url usecase

igorissen committed:
refactor(api): remove updateLastLoggedAt method

from user-repository

igorissen committed:
refactor(api): remove lasLoggedAt property in buildUser factory

igorissen committed:
refactor(api): modify getUserDetailsForAdmin method

from user-repository to retrieve lastLoggedAt from user-logins table

igorissen committed:
refactor(api): modify 2 methods from organization-learner-repository

in order to retrieve the lastLoggedAt value from user-logins table instead of users table

igorissen committed:
refactor(api): remove unnecessary property lastLoggedAt

only on user model

igorissen committed:
feat(api): add new seeds

Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: Clément Latzarus <clement.latzarus@pix.fr>

clemlatz commented on api/db/seeds/data/team-acces/build-users.js:
```diff
@@ -36,7 +36,25 @@ function _buildUserWithPoleEmploiAuthenticationMethod(databaseBuilder) {
+    username: 'gaal.dornick',
+    email: 'gaal.dornick@foundation.verse',
+  });
+  databaseBuilder.factory.buildUserLogin({ userId: userWithLastLoggedAt.id, lastLoggedAt: new Date() });
```
**suggestion**: il faudrait mettre une date fixe ici plutôt que `new Date()`, en l'état ça rend compliqué de tester que la date a été mise à jour, car ce sera toujours la date du jour (ou en tout cas du jour où a joué les seeds).

lego-technix replied:
J'ai modifié le dernier commit sur les seeds pour mettre la date `1970-01-01`

clemlatz reviewed (approved):
\+ tests fonctionnels ✅ 

![hari approves](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExa2x3ZmRuZGhwcXAzNnl0dGNmM3pxdjM3dGdqYXVsaWdtYTJydTlseSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/x3prNvtrohRJpEAH5x/giphy.gif)

lego-technix reviewed (approved):
✅ Testé fonctionnellement avec succès avec Firefox

lego-technix commented:
Note : Attention ne toujours pas merger cette PR avant d'avoir exécuté le script `node scripts/team-acces/sync-user-logins-last-logged-at.js` de #6958 en production (à un moment où il y a peu de trafic tel que vu avec les captains).
